### PR TITLE
Sp 519 add contact arrangements

### PR DESCRIPTION
**Pull request by Coupey**

### CODE REVIEW

File: app/activeplan/homepage/contactArrangements/get.controller.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+const { getCommentText } = require('../../../../common/utils/getCommentText')
+
+const getContactArrangements = async (
+  { path, errors, errorSummary, body, renderInfo, params: { planId }, headers: { 'x-auth-token': token } },
+  res
+) => {
+  const renderDetails = renderInfo || {}
+  renderDetails.backurl = `${path.substring(0, path.lastIndexOf('/'))}#contact`
+  try {
+    const comments = await getSentencePlanComments(planId, token)
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/get.controller.js]: this is only required for the else condition, so only retrieve it then. You can get contextual errors from that call too, by sticking a catch on the await before you re-throw for the outer catch.

File: app/activeplan/homepage/contactArrangements/get.controller.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    if (body.contactArrangements !== undefined) {
+      renderDetails.contactArrangements = body.contactArrangements
+    } else {
+      renderDetails.contactArrangements = getCommentText(comments, 'LIAISON_ARRANGEMENTS')
+    }
+    return res.render(`${__dirname}/index`, { ...body, errors, errorSummary, ...renderDetails })
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/get.controller.js]: Its a matter of opinion, but for me the renderDetails object just obfuscates for me; its a black box where things get popped and then all tipped out at the end. I think it'd be clearer to define it all in the return or together elsewhere?
Maybe shove the getComments and getCommentsText in a function?
```
const contactArrangements = body.contactArrangements !== undefined ? 
  body.contactArrangements : await getPersistedComments(planId, token)
return res.render(`${__dirname}/index`, { 
  ...body, 
  errors, 
  errorSummary, 
  ...renderInfo,
  backurl,
  contactArrangements,
   })
```

Comment (jonhindle) [reply app/activeplan/homepage/contactArrangements/get.controller.js]: 🤔 or...
```
const contactArrangements = body.contactArrangements
if( contactArrangements === undefined ){
  const comments = await getSentencePlanComments(planId, token)
  renderDetails.contactArrangements = getCommentText(comments, 'LIAISON_ARRANGEMENTS')
}

File: app/activeplan/homepage/contactArrangements/get.controller.test.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    errorSummary: {},
+    contactArrangements: null,
+  }
+  const res = {
+    render: jest.fn(),
+  }
+
+  beforeEach(() => {
+    req.renderInfo = {}
+    delete req.body.contactArrangements
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/get.controller.test.js]: delete action should be done on an 'afterEach'

File: app/activeplan/homepage/contactArrangements/get.controller.test.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  })
+
+  beforeEach(() => {
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/get.controller.test.js]: two beforeEach functions.

File: app/activeplan/homepage/contactArrangements/get.controller.test.js (hunk 3 above)
Comment (jonhindle) [reply app/activeplan/homepage/contactArrangements/get.controller.test.js]: Or re-declare the req object in the beforeEach every time? That would ensure the integrity. Then renderInfo could just be in the declaration.

File: app/activeplan/homepage/contactArrangements/index.njk [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            {% if wordsOver %}
+            <span id="diversity-error" class="govuk-error-message extra-error">
+                You have {{ wordsOver }} words too many</span>
+            {% endif %}
+
+            {{ govukButton({
+                text: 'Save contact arrangements',
+                classes: 'govuk-!-margin-bottom-3'
+            }) }}
+                <p>or <a class="govuk-link" href="{{ backurl }}">cancel</a></p>
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/index.njk]: Not for this, but this is used so many times, it should be a component. Worth raising a ticket?

File: app/activeplan/homepage/contactArrangements/post.controller.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+const { getContactArrangements } = require('./get.controller')
+
+const validationRules = () => {
+  return [
+    body('contactArrangements')
+      .isLength({ min: 1 })
+      .withMessage('Enter contact arrangements'),
+    body('contactArrangements')
+      .trim()
+      .escape(),
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/post.controller.js]: body('contactArrangements', 'Enter contact arrangements')
      .isLength({ min: 1 })
      .trim()
      .escape(),

File: app/activeplan/homepage/contactArrangements/get.controller.js (hunk 2 above)
Comment (Coupey) [reply app/activeplan/homepage/contactArrangements/get.controller.js]: Yeah I agree - renderDetails always seemed a little odd. Fixed.

File: app/activeplan/homepage/contactArrangements/post.controller.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    body: { contactArrangements },
+  } = req
+  if (errors) {
+    return getContactArrangements(req, res)
+  }
+  try {
+    const comment = [
+      {
+        comment: contactArrangements,
+        commentType: 'LIAISON_ARRANGEMENTS',
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/post.controller.js]: Can you raise a ticket to make some constants please?

File: app/activeplan/homepage/contactArrangements/post.controller.test.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    path: '/this/is/my/path',
+    params: {
+      planId: 1,
+    },
+    headers: {
+      'x-auth-token': '1234',
+    },
+    body: {},
+  }
+  setSentencePlanComment.mockReset()
```
Comment (jonhindle) [inline app/activeplan/homepage/contactArrangements/post.controller.test.js]: I think this really should be in an `afterEach` block.

File: app/activeplan/homepage/contactArrangements/index.njk (hunk 5 above)
Comment (Coupey) [reply app/activeplan/homepage/contactArrangements/index.njk]: Not completely convinced this is component worthy, but happy to have the argument at a later date!

File: app/activeplan/homepage/contactArrangements/post.controller.js (hunk 6 above)
Comment (Coupey) [reply app/activeplan/homepage/contactArrangements/post.controller.js]: I quite like keeping the validation and sanitisation separate but no biggy. Changed.

File: app/activeplan/homepage/contactArrangements/post.controller.js (hunk 7 above)
Comment (Coupey) [reply app/activeplan/homepage/contactArrangements/post.controller.js]: Done - https://dsdmoj.atlassian.net/browse/SP-560

File: app/objectiveReview/get.controller.test.js [hunk 9]
```diff
@@ -181,7 +181,7 @@ describe('getObjectiveReview', () => {
         )
       })
     })
-    describe('errors being thrown when there are data erros', () => {
+    describe('errors being thrown when there are data errors', () => {
```
Comment (jonhindle) [inline app/objectiveReview/get.controller.test.js]: 🤣

File: app/router.js [hunk 10]
```diff
@@ -105,6 +110,15 @@ module.exports = app => {
   app.get(`${offenderRoute}/edit-plan/new`, createSentencePlan)
   app.get(editPlanRoute, editPlan)
 
+  // edit contact arrangements
+  app.get(`${activePlanRoute}/contact-arrangements`, getContactArrangements)
+  app.post(
+    `${activePlanRoute}/contact-arrangements`,
```
Comment (jonhindle) [inline app/router.js]: Duplication `${activePlanRoute}/contact-arrangements` but I'll merge and we can clean it up in housekeeping.

**Pull request merged**